### ISSUE TITLE
Second order updates

### ISSUE THREAD

**Issue by BabaYara** [patrick-kidger/equinox#546]
Hello Patrick,

I hope you're doing well. I've been working with the Equinox package, and I appreciate the effort that's gone into it.

Currently, I'm working on implementing a second-order Taylor optimization of neural networks. A vital part of this implementation requires computing both the gradients and Hessians of the flattened parameters with respect to the loss function. This will allow me to define the updates for the model more effectively as below.

Here's a basic outline of what I'm trying to achieve:

`
def loss(params, shapes, input_data, labels):  
    return loss_0(params, shapes, input_data, labels)

grad_f = jax.grad(loss)
hessian_f = jax.hessian(loss)
lr = 0.9 

grad_params = grad_f(params, shapes, input_data, labels)
hessian_params = hessian_f(params, shapes, input_data, labels)       
delta_x = solve(hessian_params, grad_params)
params += (lr * delta_x)
` 
Could you please guide me on whether Equinox supports this kind of operation or if there's a recommended approach within the broader Jax ecosystem to achieve this?

Thanks in advance for your assistance.

**Comment by lockwo** [comment]
I don't see why there wouldn't be support, you can construct the hessian from `vmap`, `vjp`, and `jvp` all of which are supported by equinox (https://docs.kidger.site/equinox/api/filtering/transformations/#equinox.filter_jvp).

**Comment by BabaYara** [comment]
Any chance I can please get a simple working example? 
If not, I will try and get one working through the course of the week and reply to this thread so others can piggyback off it.

**Comment by lockwo** [comment]
Here is a simple example, probably not optimized for speed, but just a quick draft to help you get started in a direction.

```python
import jax
import equinox as eqx
from jax import numpy as jnp
import matplotlib.pyplot as plt
from jax import flatten_util

key = jax.random.PRNGKey(42)
key, subkey1, subkey2 = jax.random.split(key, 3)
data = jnp.concatenate((jax.random.normal(subkey1, shape=(100, 2)) * 0.1 - 1, jax.random.normal(subkey2, shape=(100, 2)) * 0.1 + 1), axis=0)
labels = jnp.array([0] * 100 + [1] * 100)
plt.scatter(data[:,0], data[:,1])
plt.show()

class MLP(eqx.Module):
    layers: list

    def __init__(self, key):
        key1, key2, key3 = jax.random.split(key, 3)
        self.layers = [
            eqx.nn.Linear(2, 10, key=key1),
            jax.nn.relu,
            eqx.nn.Linear(10, 12, key=key2),
            jax.nn.relu,
            eqx.nn.Linear(12, 2, key=key3),
            jax.nn.log_softmax
        ]

    def __call__(self, x):
        for layer in self.layers:
            x = layer(x)
        return x

def loss_fn(model, ins, ytrue):
    pred_y = eqx.filter_vmap(model)(ins)
    return cross_entropy(ytrue, pred_y)

def cross_entropy(y, pred_y):
    pred_y = jnp.take_along_axis(pred_y, jnp.expand_dims(y, 1), axis=1)
    return -jnp.mean(pred_y)

@eqx.filter_jit
def compute_accuracy(m, x, y):
    pred_y = eqx.filter_vmap(m)(x)
    pred_y = jnp.argmax(pred_y, axis=1)
    return jnp.mean(y == pred_y)

key = jax.random.PRNGKey(42)

@eqx.filter_jit
def step(mlp, xs, ys):
    vals, grads = eqx.filter_value_and_grad(loss_fn)(mlp, xs, ys)
    updates = jax.tree_map(lambda g: -0.1 * g, grads)
    mlp = eqx.apply_updates(mlp, updates)
    return mlp, vals

epochs = 50
batch_size = 100
grad_loss = []
grad_acc = []

key, subkey = jax.random.split(key, 2)
model = MLP(subkey)

for e in range(epochs):
    if e % 20 == 0:
        print(e, "/", epochs)
    
    key, subkey = jax.random.split(key, 2)
    inds = jax.random.randint(subkey, minval=0, maxval=len(data), shape=(batch_size,))
    inputs = data[inds]
    ls = labels[inds]
    
    model, loss = step(model, inputs, ls)
    grad_loss.append(loss)
    grad_acc.append(compute_accuracy(model, data, labels))

@eqx.filter_jit
def loss_h(arrs, static, ins, ytrue, uf):
    arrs = uf(arrs)
    model = eqx.combine(arrs, static)
    pred_y = eqx.filter_vmap(model)(ins)
    return cross_entropy(ytrue, pred_y)

@eqx.filter_jit
def step_h(mlp, xs, ys):
    vals, grads = eqx.filter_value_and_grad(loss_fn)(mlp, xs, ys)
    a, s = eqx.partition(mlp, eqx.is_inexact_array)
    flat_a, unflat_a = flatten_util.ravel_pytree(a)
    h = jax.hessian(loss_h)(flat_a, s, xs, ys, unflat_a)
    g_flat, unflat = flatten_util.ravel_pytree(grads)
    updates = unflat(-1 * jnp.linalg.pinv(h) @ g_flat)
    mlp = eqx.apply_updates(mlp, updates)
    return mlp, vals

key = jax.random.PRNGKey(1)
key, subkey = jax.random.split(key, 2)
model_h = MLP(subkey)

epochs = 50
batch_size = 100
h_loss = []
h_acc = []

for e in range(epochs):
    if e % 20 == 0:
        print(e, "/", epochs)
    
    key, subkey = jax.random.split(key, 2)
    inds = jax.random.randint(subkey, minval=0, maxval=len(data), shape=(batch_size,))
    inputs = data[inds]
    ls = labels[inds]
    
    model_h, loss = step_h(model_h, inputs, ls)
    h_loss.append(loss_fn(model_h, data, labels))
    h_acc.append(compute_accuracy(model_h, data, labels))

fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 4))

ax1.plot(h_loss, label="Newton")
ax2.plot(h_acc, label="Newton")

ax1.plot(grad_loss, label="Grad")
ax2.plot(grad_acc, label="Grad")

plt.legend()
plt.show()

```


Yields
<img width="382" alt="Screenshot 2023-10-08 at 11 37 59 PM" src="https://github.com/patrick-kidger/equinox/assets/42878312/8adb5654-7469-4c3b-be7d-b8f5f77167f7">

<img width="596" alt="Screenshot 2023-10-08 at 11 38 13 PM" src="https://github.com/patrick-kidger/equinox/assets/42878312/4b28c9ae-2421-4436-bba7-557d017e9865">

**Comment by BabaYara** [comment]
Thanks a lot for sharing this. Made my life so much easier.

**Comment by lockwo** [comment]
Indeed, that's all my example does, partitions then uses `jax.hessian` then combines (there is also a flattening operation just to show the matrix inversion).

**Comment by patrick-kidger** [comment]
Great!
With respect to the original question, by the way -- on a related note, you might like [Optimistix](https://github.com/patrick-kidger/optimistix), which includes some second-order optimisers.

**Comment by BabaYara** [comment]
Thanks a lot for all the offers of help here.
I took a look at Optimistix over the weekend, but will dig deeper. 
The examples come in handy a lot for people like myself so as many examples as we can put in the documentation the better for folks like me.
